File: 301. Remove Invalid Parentheses.py
class Solution:
    def removeInvalidParentheses(self, s: str) -> List[str]:
        '''
        Intuition is we count how many to remove,
        and then use dfs to remove those need to remove
        '''
        left, right = 0,0
        for char in s:
            if char == '(':
                left = left + 1 
            elif char == ')':
                if not left:
                    right += 1
                else:
                    left -= 1
        
        ans = set()
        
        
        def recursion(idx, left_count, right_count, left_rem, right_rem, cur):
            if idx == len(s):
                if left_count == right_count and (not left_rem and not right_rem):
                    ans.add(''.join(cur))
            else:
                if s[idx] == '(':
                    if left_rem:
                        recursion(idx+1, left_count, right_count, left_rem-1, right_rem, cur)
                    recursion(idx+1, left_count+1, right_count, left_rem, right_rem, cur+[s[idx]])
                elif s[idx] == ')':
                    if right_rem:
                        recursion(idx+1, left_count, right_count, left_rem, right_rem-1, cur)
                    if right_count < left_count:
                        recursion(idx+1, left_count, right_count+1, left_rem, right_rem, cur+[s[idx]])
                else:
                    recursion(idx+1, left_count, right_count, left_rem, right_rem, cur+[s[idx]])
        
        recursion(0, 0, 0, left, right, [])
        return list(ans)

File: 301. Remove Invalid Parentheses.java
/**
 * key point is find out how many to remove and we need to go through every situation
 * Time O(2^n) since worstly we may go through every char to find out whether include it or leave it
 * Space O(n) since stack will go as deep as n
 */
class Solution {
    public List<String> removeInvalidParentheses(String s) {
        Set<String> ans = new HashSet<>();
        int left = 0, right = 0;
        for (int i=0; i<s.length(); i++) {
            if (s.charAt(i) == '(') {
                left++;
            } else if (s.charAt(i) == ')') {
                right = left == 0? right+1:right;
                left = left == 0? left: left-1;
            }
        }
        helper(s, 0, 0, 0, left, right, new StringBuilder(), ans);
        return new ArrayList<>(ans);
    }
    private void helper(String s, int index, int leftCount, int rightCount, 
                       int leftRem, int rightRem, StringBuilder exp,
                        Set<String> ans) {
        if (index == s.length()) {
            // this is a valid combination
            if (leftRem == 0 && rightRem == 0) {
                ans.add(exp.toString());
            }
        } else {
            char c = s.charAt(index);
            int len = exp.length();
            // we skip this character by removing the ( or )
            if ((c == '(' && leftRem > 0) || (c == ')' && rightRem > 0)) {
                helper(s, index+1, leftCount, rightCount, 
                      leftRem - (c == '(' ? 1 : 0), 
                      rightRem - (c == ')' ? 1 : 0),
                      exp, ans);
            }
            exp.append(c);
            // skip non ( ) character
            if (c != '(' && c != ')') {
                helper(s, index+1, leftCount, rightCount, leftRem, rightRem, exp, ans);
            } else if (c == '(') {
                // we count this as valid and goes further
                helper(s, index+1, leftCount+1, rightCount, leftRem, rightRem, exp, ans);
            } else if (rightCount < leftCount) {
                // we count this as valid and goes further
                helper(s, index+1, leftCount, rightCount+1, leftRem, rightRem, exp, ans);
            }
            // backtracking
            exp.deleteCharAt(len);   
        }
    }
}